<a href="https://colab.research.google.com/github/omega-u20/SmartCare-HospitalManagement/blob/main/Task2_Dataset_Understanding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task 02 – Dataset Understanding


Connecting to Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

DATA_FOLDER = '/content/drive/MyDrive/CCS3440 - Asg 2'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## 1. Setup

In [ ]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

df = pd.read_csv(f'{DATA_FOLDER}/smartcare_ai_dataset_1000.csv')
data_dict = pd.read_csv(f'{DATA_FOLDER}/smartcare_ai_dataset_data_dictionary.csv')

print('Dataset shape:', df.shape)
df.head()

Dataset shape: (1000, 33)


,record_id,patient_id,age,gender,blood_group,department,diagnosis,appointment_date,waiting_days,previous_appointments,missed_previous_appointments,appointment_status,admitted,room_type,length_of_stay_days,previous_admissions,systolic_bp,diastolic_bp,blood_sugar_mg_dl,cholesterol_mg_dl,bmi,lab_tests_count,treatments_count,consultation_fee_lkr,room_charge_lkr,lab_charge_lkr,medicine_charge_lkr,total_bill_lkr,payment_status,payment_method,no_show,readmitted_30_days,disease_risk_level
0,1,P10001,53,Male,A-,General Medicine,Migraine,2025-04-10,10,1,0,Completed,0,NaN,0,1,127,75,117,211,26.1,0,3,2000,0,0,11596,13596,Paid,Insurance,0,0,High
1,2,P10002,26,Male,B-,General Medicine,Diabetes,2025-05-15,2,3,1,Completed,0,NaN,0,0,130,73,136,173,32.8,0,1,2000,0,0,3652,5652,Paid,Insurance,0,0,Medium
2,3,P10003,22,Male,B+,Orthopedics,Back Pain,2025-07-09,22,7,1,No-Show,0,NaN,0,1,141,64,90,176,29.4,1,0,2500,0,1200,2562,6262,Unpaid,Insurance,1,0,Medium
3,4,P10004,44,Female,AB-,Cardiology,Asthma,2025-10-16,16,1,0,Completed,0,NaN,0,0,124,82,126,189,24.9,2,1,2000,0,5000,10262,17262,Paid,Online,0,0,Medium
4,5,P10005,51,Female,O+,Neurology,Hypertension,2025-12-18,12,4,0,Scheduled,0,NaN,0,1,119,81,65,195,27.0,2,0,4000,0,6000,10414,20414,Paid,Cash,0,0,Medium


## 2. Dataset Overview

The dataset contains **1000 patient records** across **33 columns**. Each row represents one hospital appointment/visit, combining patient demographics, clinical measurements, hospital operations and billing information.

The dataset actually supports three different prediction problems (no-show, readmission, disease risk) so it includes three target columns. For this coursework we are only using `readmitted_30_days` as the target.

In [ ]:
print('Number of rows:', df.shape[0])
print('Number of columns:', df.shape[1])
print()
print('Column data types:')
print(df.dtypes.value_counts())

Number of rows: 1000
Number of columns: 33

Column data types:
int64      21
object     11
float64     1
Name: count, dtype: int64


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     1000 non-null   int64  
 1   patient_id                    1000 non-null   object 
 2   age                           1000 non-null   int64  
 3   gender                        1000 non-null   object 
 4   blood_group                   1000 non-null   object 
 5   department                    1000 non-null   object 
 6   diagnosis                     1000 non-null   object 
 7   appointment_date              1000 non-null   object 
 8   waiting_days                  1000 non-null   int64  
 9   previous_appointments         1000 non-null   int64  
 10  missed_previous_appointments  1000 non-null   int64  
 11  appointment_status            1000 non-null   object 
 12  admitted                      1000 non-null   int64  
 13  room

**Observation:** All columns are either whole numbers (`int64`), decimals (`float64`), or text (`object`). There is no obvious sign of corrupted formatting at this stage — this will be checked more carefully in the Data Quality section below.

## 3. Attribute Description

The 33 columns can be grouped into five categories:

| Category | Columns |
|---|---|
| **Identifiers** | record_id, patient_id |
| **Patient Demographics** | age, gender, blood_group |
| **Appointment / Operations** | department, diagnosis, appointment_date, waiting_days, previous_appointments, missed_previous_appointments, appointment_status, admitted, room_type, length_of_stay_days, previous_admissions |
| **Clinical Measurements** | systolic_bp, diastolic_bp, blood_sugar_mg_dl, cholesterol_mg_dl, bmi, lab_tests_count, treatments_count |
| **Financial** | consultation_fee_lkr, room_charge_lkr, lab_charge_lkr, medicine_charge_lkr, total_bill_lkr, payment_status, payment_method |
| **Target Variables** | no_show, readmitted_30_days, disease_risk_level |

Let's look at basic statistics for the numeric columns and the unique values for the categorical (text) columns.

In [ ]:
# Summary statistics for numeric columns
df.describe()

,record_id,age,waiting_days,previous_appointments,missed_previous_appointments,admitted,length_of_stay_days,previous_admissions,systolic_bp,diastolic_bp,blood_sugar_mg_dl,cholesterol_mg_dl,bmi,lab_tests_count,treatments_count,consultation_fee_lkr,room_charge_lkr,lab_charge_lkr,medicine_charge_lkr,total_bill_lkr,no_show,readmitted_30_days
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,500.500000,44.744000,21.851000,2.884000,0.553000,0.330000,1.103000,0.85500,128.423000,78.828000,114.832000,204.814000,25.519100,2.292000,1.805000,2820.500000,2412.50000,4869.600000,7971.000000,18073.600000,0.495000,0.253000
std,288.819436,17.851474,13.040892,1.693343,0.740094,0.470448,1.885191,0.95859,15.486895,10.043669,26.017007,36.645577,4.252852,1.679749,1.596723,837.694146,11140.88026,4122.515127,4629.777041,13710.258789,0.500225,0.434948
min,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,85.000000,50.000000,65.000000,100.000000,14.000000,0.000000,0.000000,1500.000000,0.00000,0.000000,506.000000,2954.000000,0.000000,0.000000
25%,250.750000,33.000000,11.000000,2.000000,0.000000,0.000000,0.000000,0.00000,117.000000,72.000000,97.000000,180.000000,22.700000,1.000000,1.000000,2000.000000,0.00000,2400.000000,4316.500000,11018.000000,0.000000,0.000000
50%,500.500000,44.000000,22.000000,3.000000,0.000000,0.000000,0.000000,1.00000,128.000000,79.000000,114.000000,204.000000,25.600000,2.000000,1.000000,3000.000000,0.00000,3600.000000,7680.000000,15068.500000,0.000000,0.000000
75%,750.250000,57.000000,34.000000,4.000000,1.000000,1.000000,2.000000,1.00000,139.000000,86.000000,132.000000,230.000000,28.300000,3.000000,3.000000,3500.000000,0.00000,6300.000000,11032.000000,20870.000000,1.000000,1.000000
max,1000.000000,90.000000,44.000000,10.000000,4.000000,1.000000,9.000000,5.00000,178.000000,111.000000,201.000000,330.000000,38.800000,10.000000,10.000000,4000.000000,150000.00000,27000.000000,20371.000000,179528.000000,1.000000,1.000000


In [ ]:
# Categorical columns and their unique values
categorical_cols = df.select_dtypes(include='object').columns.tolist()
categorical_cols.remove('patient_id')  # identifier, not a real category
categorical_cols.remove('appointment_date')  # date, not a category

for col in categorical_cols:
    print(f"--- {col} ---")
    print(df[col].value_counts())
    print()

--- gender ---
gender
Male      501
Female    499
Name: count, dtype: int64

--- blood_group ---
blood_group
AB-    136
A-     135
O-     135
B+     132
AB+    122
A+     120
B-     112
O+     108
Name: count, dtype: int64

--- department ---
department
General Medicine       159
Cardiology             158
Neurology              148
Orthopedics            141
Radiology              140
Pediatrics             138
Laboratory Services    116
Name: count, dtype: int64

--- diagnosis ---
diagnosis
Pneumonia           115
Fever               107
Back Pain           104
Migraine            104
Fracture            104
Kidney Infection    101
Diabetes             97
Asthma               92
Hypertension         89
Chest Pain           87
Name: count, dtype: int64

--- appointment_status ---
appointment_status
No-Show      495
Completed    380
Scheduled     72
Cancelled     53
Name: count, dtype: int64

--- room_type ---
room_type
General Ward    52
Private Room    33
ICU              9
Name: cou

### Key attribute notes

- **age** ranges roughly across the adult population, no negative or impossible values expected — checked in Data Quality below.
- **gender** and **blood_group** are simple categorical fields.
- **department / diagnosis** describe what the visit was for.
- **admitted** is a 0/1 flag. `room_type` and `length_of_stay_days` only make sense when `admitted = 1`, which is why we expect a lot of missing/None values in `room_type` for non-admitted patients.
- **previous_admissions** and **missed_previous_appointments** are historical counts that are likely to be strong predictors of readmission, since past behaviour tends to predict future behaviour.
- **systolic_bp, diastolic_bp, blood_sugar_mg_dl, cholesterol_mg_dl, bmi** are the core clinical vitals used to judge a patient's health condition.
- **total_bill_lkr** is effectively the sum of consultation, room, lab, and medicine charges — this will be worth checking for consistency.

## 4. Target Variable – `readmitted_30_days`

This is the column we are trying to predict for Option B. It is a **binary classification** target:
- `1` = patient was readmitted within 30 days
- `0` = patient was not readmitted within 30 days

In [ ]:
counts = df['readmitted_30_days'].value_counts()
percentages = df['readmitted_30_days'].value_counts(normalize=True) * 100

print('Class counts:')
print(counts)
print()
print('Class percentages:')
print(percentages.round(1))

Class counts:
readmitted_30_days
0    747
1    253
Name: count, dtype: int64

Class percentages:
readmitted_30_days
0    74.7
1    25.3
Name: proportion, dtype: float64


**Observation:** The dataset has 747 patients not readmitted (74.7%) and 253 patients readmitted (25.3%). This is a **moderately imbalanced** target — not extreme, but the minority class (readmitted) is about a third the size of the majority class. This is something we will need to handle carefully at the modelling stage (Task 05), for example by using class weighting, resampling, or choosing evaluation metrics like F1-score and ROC-AUC instead of relying only on accuracy.

## 5. Input Variables vs Target Variable

For the readmission prediction task:

- **Target (label):** `readmitted_30_days`
- **Candidate input features:** all other columns *except* the identifiers (`record_id`, `patient_id`) and the two unused target columns (`no_show`, `disease_risk_level`), which must be dropped so the model doesn't accidentally "cheat" by learning from a related target.

`appointment_date` is a raw date string and will need to be converted into something more useful (e.g. month, day of week) or dropped during Feature Engineering in Task 03  it cannot be fed into a model as is.

In [ ]:
target = 'readmitted_30_days'
cols_to_exclude = ['record_id', 'patient_id', 'no_show', 'disease_risk_level', target]
candidate_features = [c for c in df.columns if c not in cols_to_exclude]

print('Target variable:', target)
print()
print(f'Candidate input features ({len(candidate_features)}):')
print(candidate_features)

Target variable: readmitted_30_days

Candidate input features (28):
['age', 'gender', 'blood_group', 'department', 'diagnosis', 'appointment_date', 'waiting_days', 'previous_appointments', 'missed_previous_appointments', 'appointment_status', 'admitted', 'room_type', 'length_of_stay_days', 'previous_admissions', 'systolic_bp', 'diastolic_bp', 'blood_sugar_mg_dl', 'cholesterol_mg_dl', 'bmi', 'lab_tests_count', 'treatments_count', 'consultation_fee_lkr', 'room_charge_lkr', 'lab_charge_lkr', 'medicine_charge_lkr', 'total_bill_lkr', 'payment_status', 'payment_method']


## 6. Data Dictionary Interpretation

The dataset comes with an official data dictionary file (`smartcare_ai_dataset_data_dictionary.csv`) describing what every column means. Below we display it in full, and highlight the columns most relevant to our readmission prediction task.

In [ ]:
data_dict

,Column,Description
0,record_id,Unique row identifier
1,patient_id,Synthetic patient identifier
2,age,Patient age in years
3,gender,Patient gender
4,blood_group,Patient blood group
5,department,Hospital department
6,diagnosis,Primary diagnosis category
7,appointment_date,Appointment date
8,waiting_days,Number of days between booking and appointment
9,previous_appointments,Number of previous appointments


In [ ]:
# Highlight the dictionary rows most relevant to readmission prediction
readmission_relevant = ['previous_admissions', 'missed_previous_appointments', 'admitted',
                          'length_of_stay_days', 'room_type', 'systolic_bp', 'diastolic_bp',
                          'blood_sugar_mg_dl', 'cholesterol_mg_dl', 'bmi', 'age',
                          'lab_tests_count', 'treatments_count', 'readmitted_30_days']

data_dict[data_dict['Column'].isin(readmission_relevant)]

,Column,Description
2,age,Patient age in years
10,missed_previous_appointments,Number of previously missed appointments
12,admitted,"Whether patient was admitted: 1=Yes, 0=No"
13,room_type,"None, General Ward, Private Room, or ICU"
14,length_of_stay_days,Hospital stay duration in days
15,previous_admissions,Number of previous admissions
16,systolic_bp,Synthetic systolic blood pressure
17,diastolic_bp,Synthetic diastolic blood pressure
18,blood_sugar_mg_dl,Synthetic blood sugar measurement
19,cholesterol_mg_dl,Synthetic cholesterol measurement


**Interpretation:** These columns describe a patient's clinical severity (blood pressure, sugar, cholesterol, BMI) and their hospital history (previous admissions, length of stay, missed appointments). Together they form the most medically sensible set of predictors for whether someone comes back within 30 days — this matches how readmission risk is generally understood in healthcare: sicker patients and patients with a history of frequent hospital contact are more likely to be readmitted.

## 7. Data Quality Check

A quick first look at data quality

In [ ]:
# Missing values
missing = df.isnull().sum()
missing = missing[missing > 0]
print('Columns with missing values:')
print(missing)
print()
print(f"Percentage of rows missing 'room_type': {df['room_type'].isnull().mean()*100:.1f}%")

Columns with missing values:
room_type    906
dtype: int64

Percentage of rows missing 'room_type': 90.6%


In [ ]:
# Cross-check: is room_type missing exactly when the patient was NOT admitted?
pd.crosstab(df['admitted'], df['room_type'].isnull(), rownames=['admitted'], colnames=['room_type is missing'])

room_type is missing,False,True
admitted,,
0,0,670
1,94,236


**Observation:** `room_type` is missing for 906 out of 1000 records (90.6%). This is **not a data error** — the crosstab confirms it is missing exactly for patients where `admitted = 0`. In other words, if a patient wasn't admitted, they were never assigned a room, so `room_type` is naturally blank. This will be handled in Task 03 by filling these with a value such as `'Not Admitted'` rather than treating it as a missing-data problem.

In [ ]:
# Duplicate records
print('Fully duplicate rows:', df.duplicated().sum())
print('Duplicate patient IDs:', df['patient_id'].duplicated().sum())

Fully duplicate rows: 0
Duplicate patient IDs: 0


**Observation:** No duplicate rows and no duplicate patient IDs were found, so each record represents a unique patient visit.

In [ ]:
# Sanity check on numeric ranges
df[['age', 'bmi', 'systolic_bp', 'diastolic_bp', 'blood_sugar_mg_dl', 'cholesterol_mg_dl']].describe()

,age,bmi,systolic_bp,diastolic_bp,blood_sugar_mg_dl,cholesterol_mg_dl
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,44.744000,25.519100,128.423000,78.828000,114.832000,204.814000
std,17.851474,4.252852,15.486895,10.043669,26.017007,36.645577
min,1.000000,14.000000,85.000000,50.000000,65.000000,100.000000
25%,33.000000,22.700000,117.000000,72.000000,97.000000,180.000000
50%,44.000000,25.600000,128.000000,79.000000,114.000000,204.000000
75%,57.000000,28.300000,139.000000,86.000000,132.000000,230.000000
max,90.000000,38.800000,178.000000,111.000000,201.000000,330.000000


**Observation:** The min/max values for age, BMI, and the clinical vitals look within medically plausible ranges (no negative ages, no zero BMI, etc.). A more thorough outlier check using boxplots will be done in Task 03/04.